In [183]:
import pandas as pd

df = pd.read_csv("../data/raw/Online_Retail.csv")

In [184]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/10 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/10 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/10 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/10 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/10 8:26,3.39,17850.0,United Kingdom


In [8]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/11 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/11 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/11 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/11 12:50,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,12/9/11 12:50,4.95,12680.0,France


In [29]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [37]:
df.index

RangeIndex(start=0, stop=541909, step=1)

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [189]:
df.shape

(541909, 8)

In [146]:
df.isnull().sum()
# print(df[df['CustomerID'].isnull()])

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Findings:
Description and CustomerID have 1454 and 135080 null values respectively

Silver implications:
decide how to handle null customers
clean and review missing descriptions; don’t assume every product row has a usable description

In [196]:
df[['InvoiceNo', 'StockCode', 'Country','CustomerID','Description']].nunique()

InvoiceNo      25900
StockCode       4070
Country           38
CustomerID      4372
Description     4223
dtype: int64

In [ ]:
#7. Check date range
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

earliest_date = df['InvoiceDate'].min()
last_date = df['InvoiceDate'].max()

print(earliest_date)
print (last_date)
print()

df.info()

2010-12-01 08:26:00
2011-12-09 12:50:00

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


InvoiceDate was originally read as an object/string
•	It was successfully converted to datetime64[ns]
•	Date coverage is from 2010-12-01 08:26:00 to 2011-12-09 12:50:00
•	Silver should store invoice_date as a datetime field

In [ ]:
#8. Check cancelations
print(f"Num Rows with InvoiceNo starting with C, indicating a cancelation: {df['InvoiceNo'].str.startswith('C').sum()}")

df[df['InvoiceNo'].str.startswith('C')]



Num Rows with InvoiceNo starting with C, indicating a cancelation: 9288


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


findings:
9,288 rows where InvoiceNo starts with C

Silver implication: 
create a cancellation flag, probably based on InvoiceNo.str.startswith('C')

In [120]:
#9. Check negative quanity counts
print(f"Num Rows with Negative quantity counts: {(df['Quantity'] < 0).sum()}")

df[df['Quantity'] < 0]

Num Rows with Negative quantity counts: 10624


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


In [203]:
print(f"Num Rows with InvoiceNo starting with C and negative quanity counts: {((df['InvoiceNo'].str.startswith('C')) & (df['Quantity'] < 0)).sum()}")

df[(df['InvoiceNo'].str.startswith('C')) & (df['Quantity'] < 0)]

Num Rows with InvoiceNo starting with C and negative quanity counts: 9288


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/10 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/10 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/10 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/10 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/10 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/11 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/11 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/11 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/11 11:58,1.25,17315.0,United Kingdom


In [204]:
print(f"Num Rows with InvoiceNo starting with C and negative quanity counts: {((~df['InvoiceNo'].str.startswith('C')) & (df['Quantity'] < 0)).sum()}")

df[(~df['InvoiceNo'].str.startswith('C')) & (df['Quantity'] < 0)]

Num Rows with InvoiceNo starting with C and negative quanity counts: 1336


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/10 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/10 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/10 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/10 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/10 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/11 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/11 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/11 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/11 9:56,0.0,NaN,United Kingdom


findings:
	•	10,624 rows with negative quantity
	•	but only 9,288 cancellation rows

So there are 1,336 negative-quantity rows beyond the C invoices.

Silver implication: 
keep separate logic/flags for:
	•	cancellation invoice
	•	negative quantity

Finding:
	There are 10,624 rows with negative quantity.
	•	There are 9,288 rows where InvoiceNo begins with C.
	•	This means negative quantity and cancellation invoices overlap heavily but are not exactly the same.
	•	Negative-quantity rows also contain missing Description and CustomerID values.

In [187]:
# Check negative prices
print(f"Num of rows with Unit price < 0: {(df['UnitPrice'] < 0.0).sum()}")

print(df[df['UnitPrice'] < 0.0])




Num of rows with Unit price < 0: 2
       InvoiceNo StockCode      Description  Quantity    InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/11 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/11 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom  



findings:
These are almost certainly special accounting/adjustment records, not normal sales.

Silver implication:
These will be:
	•flagged as invalid sales rows
	•excluded from the core silver sales table, depending on your design

these will not be treated as normal customer purchase transactions.

In [186]:
print(f"Num of rows with Unit price == 0: {(df['UnitPrice'] == 0.0).sum()}")

print(df[df['UnitPrice'] == 0.0])

Num of rows with Unit price == 0: 2515
       InvoiceNo StockCode                   Description  Quantity  \
622       536414     22139                           NaN        56   
1970      536545     21134                           NaN         1   
1971      536546     22145                           NaN         1   
1972      536547     37509                           NaN         1   
1987      536549    85226A                           NaN         1   
...          ...       ...                           ...       ...   
536981    581234     72817                           NaN        27   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538554    581408     85175                           NaN        20   
538919    581422     23169                       smashed      -235   

          InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/10 11:52        0.0         NaN  United King

findings:
many more rows with UnitPrice == 0 than UnitPrice < 0
Also good to note that there are rows with NaN customerID and NaN Description (need to find later)

silver implications:
They may represent:
	•	freebies
	•	samples
	•	promotional items
	•	placeholders
	•	incomplete records
	•	nonstandard operational rows

    will not lump UnitPrice == 0 together with UnitPrice < 0 too aggressively.

In [ ]:
#10. Check rows with empty descriptions
print(f"Num of rows with empty descriptions (NaN): {(df['Description'].isnull().sum())}")

df[df['Description'].isnull()]

Num of rows with empty descriptions (NaN): 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


findings:
These do not look like clean retail sales lines.
They look like incomplete, nonstandard, or placeholder/adjustment-style records.

Silver implication:
- keep rows with missing Description only if other transaction fields suggest the row is still analytically meaningful
	•	otherwise flag them as nonstandard / low-trust rows
	•	gold sales marts may exclude rows lacking usable product attributes

Reasonable options:
	•keep them in bronze only

In [185]:
print(f"Num of rows with empty CustomerIDs (NaN): {(df['CustomerID'].isnull().sum())}")

df[df['CustomerID'].isnull()]

Num of rows with empty CustomerIDs (NaN): 135080


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/10 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/10 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/10 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/10 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/10 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/11 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/11 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/11 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/11 10:26,10.79,NaN,United Kingdom


findings: 
lots of CustomerID that are null but otherwise have valid invoice orders

silver implication:
keep rows with null CustomerID in silver if they are otherwise valid transaction records
	•	do not let null CustomerID break the pipeline
	•	decide in gold whether certain customer-focused marts exclude unknown customers


In [192]:
print(f"Number of correlating UnitPrice == 0 and Description having 'NaN' and CustomerID having 'NaN': {((df['UnitPrice'] == 0) & (df['Description'].isnull()) & (df['CustomerID'].isnull())).sum()}")

df[(df['UnitPrice'] == 0) & (df['Description'].isnull()) & (df['CustomerID'].isnull())]


Number of correlating UnitPrice == 0 and Description having 'NaN' and CustomerID having 'NaN': 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/10 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/10 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/10 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/11 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/11 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/11 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/11 10:33,0.0,NaN,United Kingdom


In [194]:
print(f"Number of correlating UnitPrice == 0 and Description having 'NaN': {((df['UnitPrice'] == 0) & (df['Description'].isnull())).sum()}")

df[(df['UnitPrice'] == 0) & (df['Description'].isnull())]

Number of correlating UnitPrice == 0 and Description having 'NaN': 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/10 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/10 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/10 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/11 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/11 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/11 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/11 10:33,0.0,NaN,United Kingdom


In [195]:
print(f"Number of correlating UnitPrice == 0 and CustomerID having 'NaN': {((df['UnitPrice'] == 0) & (df['CustomerID'].isnull())).sum()}")

df[(df['UnitPrice'] == 0) & (df['CustomerID'].isnull())]

Number of correlating UnitPrice == 0 and CustomerID having 'NaN': 2475


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/10 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/10 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/10 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,12/8/11 10:33,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/11 13:58,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,12/8/11 13:58,0.0,NaN,United Kingdom
538554,581408,85175,NaN,20,12/8/11 14:06,0.0,NaN,United Kingdom


In [199]:
print(f"Number of correlating Description having NaN and CustomerID having 'NaN': {((df['Description'].isnull()) & (df['CustomerID'].isnull())).sum()}")

df[(df['Description'].isnull()) & (df['CustomerID'].isnull())]

Number of correlating Description having NaN and CustomerID having 'NaN': 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/10 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/10 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/10 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/10 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/11 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/11 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/11 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/11 10:33,0.0,NaN,United Kingdom


In [197]:
print(f"Number of correlating UnitPrice == 0 and Description NOT having 'NaN' and CustomerID having 'NaN': {((df['UnitPrice'] == 0) & (df['Description'].notnull()) & (df['CustomerID'].isnull())).sum()}")

df[(df['UnitPrice'] == 0) & (df['Description'].notnull()) & (df['CustomerID'].isnull())]

Number of correlating UnitPrice == 0 and Description NOT having 'NaN' and CustomerID having 'NaN': 1021


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,12/3/10 12:08,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,12/3/10 12:08,0.0,NaN,United Kingdom
7313,537032,21275,?,-30,12/3/10 16:50,0.0,NaN,United Kingdom
13217,537425,84968F,check,-20,12/6/10 15:35,0.0,NaN,United Kingdom
13218,537426,84968E,check,-35,12/6/10 15:36,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535336,581213,22576,check,-30,12/7/11 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/11 9:56,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/11 13:58,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,12/8/11 13:58,0.0,NaN,United Kingdom


Findings:
	•	All 1,454 rows with null Description also have null CustomerID.
	•	These same rows also have UnitPrice == 0.
	•	This indicates that missing-description rows form a specific low-trust subgroup rather than ordinary incomplete sales lines.
	•	These rows likely represent incomplete, placeholder, or nonstandard operational records.

Silver implication

This gives you a cleaner rule:
	•	preserve these rows in bronze
	•	flag them in silver as nonstandard records
	•	exclude them from the core sales model used for gold analytics

Even stronger conclusion

Now your hierarchy is clearer:

Most suspicious subgroup
	•	Description null
	•	CustomerID null
	•	UnitPrice == 0

Count: 1,454

Broader suspicious subgroup
	•	UnitPrice == 0
	•	CustomerID null
	•	description may or may not exist

Count: 2,475

Less clear subgroup
	•	UnitPrice == 0
	•	description present
	•	CustomerID null

Count: 1,021


In [ ]:
#11. Check duplicates

print(f"Number of duplicate rows: {df.duplicated().sum()}") #just to check rows that have duplicates

print(df[df.duplicated(keep=False)]) #just to check rows that have duplicates

# print(df)

cleaned_df = df.drop_duplicates() #drops duplicate rows with keep='first' by default

# print(cleaned_df)


print(f"Number of duplicate rows after dropping duplicates: {cleaned_df.duplicated().sum()}")





Number of duplicate rows: 5268
Number of duplicate rows: 0


findings:
Yes, the dataset contains a meaningful number of rows that are complete copies across every column.

Silver implication:
planned rule is valid:
	•Bronze keeps them as-is
	•Silver removes exact duplicate rows while keeping one copy


Final profiling conclusions:
	•	Raw grain is one invoice line item per product.
	•	InvoiceDate is initially stored as text and should be converted to a proper datetime type.
	•	CustomerID contains many nulls and cannot be assumed to be present for every transaction row.
	•	Description contains nulls, and all null-description rows also have null CustomerID and UnitPrice == 0, indicating a specific low-trust subgroup 	 rather than random missingness.
	•	Cancellation rows are identified by InvoiceNo values beginning with C.
	•	Negative-quantity rows overlap strongly with cancellation rows, but they are not exactly the same population, so they should be treated as related but separate business conditions.
	•	There are a small number of negative-price rows that appear to be accounting or adjustment-style records rather than normal customer sales.
	•	There are many zero-price rows, and a large portion of them also have null CustomerID; these rows likely include a mix of incomplete, promotional, free-item, and other nonstandard operational records.
	•	The combination of UnitPrice == 0 + null Description + null CustomerID forms a distinct suspicious subgroup that should not be treated as normal retail sales.
	•	Exact duplicate rows exist across all columns and should be preserved in bronze but deduplicated in silver while keeping one copy.
	•	The source does not appear to contain a separate invoice-line surrogate key, so invoice-line grain must be inferred from the available business columns.
	•	The dataset supports deriving customer, product, country, and date dimensions, with a sales fact table at invoice-line grain.

Final silver implications:
	-   rename columns to snake_case
	•	convert InvoiceDate to datetime
	•	preserve invoice-line grain
	•	create is_cancellation_invoice
	•	create is_negative_quantity
	•	create is_zero_price
	•	create is_negative_price
	•	remove exact duplicate rows while keeping one copy
	•	derive line_amount
	•	retain null CustomerID rows in silver unless otherwise invalid
	•	flag or isolate nonstandard rows such as adjustment-like records
	•	decide which flagged rows are excluded from gold sales marts